common/environment/environment.cpp

```diff
@@ -1606,25 +1598,48 @@ IConstDropZoneInfo * CLocalEnvironment::getDropZoneByAddressPath(const char * ne
         addPathSepChar(fullDropZoneDir);
         const char * pdropzoneDir = fullDropZoneDir.str();
 
-        // Check target file path starts with this Drop zone path
-        // the drop zone paths can be nested (nothing forbids it) like
-        // dz1: /a/b/c/d
-        // dz2: /a/b/c
-        // dz3: /a/b
-        // check all and select the shortest
-
         if (strncmp(pdropzoneDir, targetFilePath, strlen(pdropzoneDir)) == 0)
         {
-            if (dropzonePathLen > strlen(pdropzoneDir))
+            dropZone = &zoneIt->query();
+
+            // For backward compatibility
+            SCMStringBuffer dropZoneMachineName;
+            // Returns dropzone '@server' value if it is exists
+            dropZone->getComputerName(dropZoneMachineName);
+
+            // check if it is local machine
+            if (strcmp(netaddress, dropZoneMachineName.str()) !=  0)
             {
-                dropzonePathLen = strlen(pdropzoneDir);
-                dropzonePath.set(pdropzoneDir);
-                pdropzonePath = dropzonePath.str();
-                retVal = &zoneIt->query();
+                // It isn't local machine or it has not '@server' attrib. Check the ServerList
+                bool isServerFound = false;
+                Owned<IPropertyTreeIterator> whiteListIt = dropZone->getPTree().getElements("ServerList");
+                ForEach(*whiteListIt)
+                {
+                    const char *server = whiteListIt->query().queryProp("@server");
+                    // It can be a hostname or an IP -> get the IP
+                    IpAddress serverIP(server);
+                    StringBuffer serverIpString;
+                    serverIP.getIpText(serverIpString);
+#ifdef _DEBUG
+                    LOG(MCdebugInfo, unknownJob, "Listed server: '%s', IP: '%s'", server, serverIpString.str());
+#endif
+                    if (strcmp(netaddress,serverIpString.str()) ==  0)
```

Comment:
Better to use IpAddress.ipequals() - i.e. make another IpAddress for netaaddress